File: distribution/python/dqops/client/models/dqo_queue_job_id.py
from typing import Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..types import UNSET, Unset

T = TypeVar("T", bound="DqoQueueJobId")


@_attrs_define
class DqoQueueJobId:
    """Identifies a single job that was pushed to the job queue.

    Attributes:
        job_id (Union[Unset, int]): Job id.
        job_business_key (Union[Unset, str]): Optional job business key that was assigned to the job. A business key is
            an alternative user assigned unique job identifier used to find the status of a job finding it by the business
            key.
        parent_job_id (Union[Unset, DqoQueueJobId]): Identifies a single job that was pushed to the job queue.
        created_at (Union[Unset, int]): The timestamp when the job was created.
    """

    job_id: Union[Unset, int] = UNSET
    job_business_key: Union[Unset, str] = UNSET
    parent_job_id: Union[Unset, "DqoQueueJobId"] = UNSET
    created_at: Union[Unset, int] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        job_id = self.job_id
        job_business_key = self.job_business_key
        parent_job_id: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.parent_job_id, Unset):
            parent_job_id = self.parent_job_id.to_dict()

        created_at = self.created_at

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if job_id is not UNSET:
            field_dict["jobId"] = job_id
        if job_business_key is not UNSET:
            field_dict["jobBusinessKey"] = job_business_key
        if parent_job_id is not UNSET:
            field_dict["parentJobId"] = parent_job_id
        if created_at is not UNSET:
            field_dict["createdAt"] = created_at

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        d = src_dict.copy()
        job_id = d.pop("jobId", UNSET)

        job_business_key = d.pop("jobBusinessKey", UNSET)

        _parent_job_id = d.pop("parentJobId", UNSET)
        parent_job_id: Union[Unset, DqoQueueJobId]
        if isinstance(_parent_job_id, Unset):
            parent_job_id = UNSET
        else:
            parent_job_id = DqoQueueJobId.from_dict(_parent_job_id)

        created_at = d.pop("createdAt", UNSET)

        dqo_queue_job_id = cls(
            job_id=job_id,
            job_business_key=job_business_key,
            parent_job_id=parent_job_id,
            created_at=created_at,
        )

        dqo_queue_job_id.additional_properties = d
        return dqo_queue_job_id

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties

File: dqops/src/main/java/com/dqops/core/jobqueue/DqoQueueJobId.java
/*
 * Copyright © 2021 DQOps (support@dqops.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.dqops.core.jobqueue;

import com.dqops.utils.docs.generators.SampleLongsRegistry;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import io.swagger.annotations.ApiModel;
import org.jetbrains.annotations.NotNull;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.ZoneOffset;

/**
 * Identifies a single job.
 */
@ApiModel(value = "DqoQueueJobId", description = "Identifies a single job that was pushed to the job queue.")
@JsonInclude(JsonInclude.Include.NON_NULL)
public class DqoQueueJobId implements Comparable<DqoQueueJobId> {
    @JsonPropertyDescription("Job id.")
    private long jobId;

    @JsonPropertyDescription("Optional job business key that was assigned to the job. A business key is an alternative user assigned unique job identifier used to find the status of a job finding it by the business key.")
    private String jobBusinessKey;

    @JsonPropertyDescription("Parent job id. Filled only for nested jobs, for example a sub-job that runs data quality checks on a single table.")
    private DqoQueueJobId parentJobId;

    @JsonPropertyDescription("The timestamp when the job was created.")
    private Instant createdAt;

    /**
     * Creates a new job id.
     */
    public DqoQueueJobId() {
        this.createdAt = Instant.now();
    }

    /**
     * Creates a new job, given a job id.
     * @param jobId Job id.
     */
    public DqoQueueJobId(long jobId) {
        this();
        this.jobId = jobId;
    }

    /**
     * Returns a unique job ID  that was assigned to the job.
     * @return Unique job id.
     */
    public long getJobId() {
        return jobId;
    }

    /**
     * Sets the job id. Should be used only for deserialization.
     * @param jobId Job id.
     */
    public void setJobId(long jobId) {
        this.jobId = jobId;
    }

    /**
     * Returns an optional business key that is an alternative, user assigned unique job identifier.
     * @return Job business key.
     */
    public String getJobBusinessKey() {
        return jobBusinessKey;
    }

    /**
     * Sets a job business key.
     * @param jobBusinessKey Job business key.
     */
    public void setJobBusinessKey(String jobBusinessKey) {
        this.jobBusinessKey = jobBusinessKey;
    }

    /**
     * Returns the timestamp when the job was created.
     * @return Job created timestamp.
     */
    public Instant getCreatedAt() {
        return createdAt;
    }

    /**
     * Sets the created timestamp. Should be used only for deserialization.
     * @param createdAt Created at timestamp.
     */
    public void setCreatedAt(Instant createdAt) {
        this.createdAt = createdAt;
    }

    /**
     * Returns an optional ID of a parent job.
     * @return Parent job id or null, when a parent job is not present.
     */
    public DqoQueueJobId getParentJobId() {
        return parentJobId;
    }

    /**
     * Sets an job id of a parent job. Should be used only for deserialization.
     * @param parentJobId ID of a parent job.
     */
    public void setParentJobId(DqoQueueJobId parentJobId) {
        this.parentJobId = parentJobId;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        DqoQueueJobId that = (DqoQueueJobId) o;

        return jobId == that.jobId;
    }

    @Override
    public int hashCode() {
        return (int) (jobId ^ (jobId >>> 32));
    }

    /**
     * Compares this object with the specified object for order.  Returns a
     * negative integer, zero, or a positive integer as this object is less
     * than, equal to, or greater than the specified object.
     */
    @Override
    public int compareTo(@NotNull DqoQueueJobId o) {
        return Long.compare(this.jobId, o.jobId);
    }

    @Override
    public String toString() {
        return "DqoQueueJobId{" +
                "jobId=" + jobId +
                ", jobBusinessKey=" + jobBusinessKey +
                ", createdAt=" + createdAt +
                '}';
    }

    public static class DqoQueueJobIdSampleFactory implements SampleValueFactory<DqoQueueJobId> {
        @Override
        public DqoQueueJobId createSample() {
            return new DqoQueueJobId() {{
                setJobId(SampleLongsRegistry.getJobId());
                setCreatedAt(LocalDateTime.of((int)SampleLongsRegistry.getYear(),
                        (int)SampleLongsRegistry.getMonth(),
                        11,
                        13,
                        42,
                        0
                ).toInstant(ZoneOffset.UTC));
            }};
        }
    }
}
